tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/GradientScope.java

```diff
@@ -0,0 +1,145 @@
+
+  @Override
+  public ExecutionEnvironment env() {
+    return graph;
+  }
+
+  @Override
+  public GradientScope withSubScope(String childScopeName) {
+    return new GradientScope(nativeScope.NewSubScope(childScopeName), graph, childScopeName);
+  }
+
+  @Override
+  public GradientScope withName(String opName) {
+    return new GradientScope(nativeScope, graph, opName);
+  }
+
+  @Override
+  public GradientScope withNameAsSubScope(String defaultName) {
+    return withSubScope(opName);
+  }
+
+  @Override
+  public GradientScope withDevice(DeviceSpec deviceSpec) {
+    return new GradientScope(nativeScope.WithDevice(deviceSpec.toString()), graph);
+  }
+
+  @Override
+  public Scope withInitScope() {
+    throw new IllegalStateException("Can't add init operations in a gradient scope");
+  }
+
+  @Override
+  public String makeOpName(String defaultName) {
+    String name = opName != null ? opName : defaultName;
+    return nativeScope.GetUniqueNameForOp(name);
+  }
+
+  @Override
+  public String makeUnique(String id) {
+    return nativeScope.GetUniqueNameForOp(id);
+  }
+
+  @Override
+  public void refreshNames() {}
+
+  @Override
+  public GradientScope withControlDependencies(Iterable<Op> controls) {
+    List<Op> controlDeps =
+        StreamSupport.stream(controls.spliterator(), false).collect(Collectors.toList());
+    NativeOperation ops = new NativeOperation(controlDeps.size());
+
+    for (int i = 0; i < controlDeps.size(); i++) {
+      Operation op = controlDeps.get(i).op();
+      if (!(op instanceof GraphOperation)) {
+        throw new IllegalArgumentException("Can only add graph ops as control dependencies");
+      }
+      ops.position(i)
+          .put(new NativeOperation(((GraphOperation) op).getUnsafeNativeHandle().node()));
+    }
+
+    return new GradientScope(nativeScope.WithControlDependencies(new NativeOperation(ops)), graph);
+  }
+
+  @Override
+  public Scope withControlDependencyOps(Iterable<Operation> controls) {
+    List<Operation> controlDeps =
+        StreamSupport.stream(controls.spliterator(), false).collect(Collectors.toList());
+    NativeOperation ops = new NativeOperation(controlDeps.size());
+
+    for (int i = 0; i < controlDeps.size(); i++) {
+      Operation op = controlDeps.get(i);
+      if (!(op instanceof GraphOperation)) {
+        throw new IllegalArgumentException("Can only add graph ops as control dependencies");
+      }
+      ops.position(i)
+          .put(new NativeOperation(((GraphOperation) op).getUnsafeNativeHandle().node()));
+    }
+
+    return new GradientScope(nativeScope.WithControlDependencies(new NativeOperation(ops)), graph);
+  }
+
+  @Override
+  public OperationBuilder apply(OperationBuilder builder) {
+    return builder;
+  }
+
+  @Override
+  public void onOpCreated(Operation op) {}
+
+  @Override
+  public String getDeviceString() {
+    throw new IllegalStateException("Can't get device string for gradient scope");
```

Comment:
`UnsupportedOperationException`? Alternatively couldn't we buffer this in Java given it has a `withDevice` method.

Reply:
My  only concern with that is that if the device is inherited from the initial graph, it won't show up, only if it's directly set.  Still worth doing though, I think.